packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -2875,9 +2879,24 @@ class _RouteEntry extends RouteTransitionRecord {
     assert(navigator._debugLocked);
     assert(route._navigator == navigator);
     currentState = _RouteLifecycle.popping;
-    navigator._observedRouteDeletions.add(
-      _NavigatorPopObservation(route, previousPresent),
-    );
+    if (route._popCompleter.isCompleted) {
+      // This is a page-based route popped through the Navigator.pop. The
+      // didPop should have been called. No further action is needed.
+      assert(hasPage);
+      assert(pendingResult == null);
+      return;
+    }
+    if (!route.didPop(pendingResult)) {
```

Comment:
before this if: `assert(!hasPage)` ? We should not be calling didPop on a page-based route right?

Reply:
the transitiondelegate not longer calls didpop on the page.
it instead will mark the page-based route for pop and will come to here

Reply:
OK, so pages that went through `Navigator.onPopPage` will always go through the first if then? Should/can we assert in line 4891 that the completer is complete to ensure that that path is choosen?